File: win-prog/ProgWinCS/Tree View and List View/ExplorerLike/ExplorerLike.cs
//-------------------------------------------
// ExplorerLike.cs © 2001 by Charles Petzold
//-------------------------------------------
using System;
using System.Drawing;
using System.Windows.Forms;

class ExplorerLike: Form
{
     FileListView      filelist;
     DirectoryTreeView dirtree;
     MenuItemView      mivChecked;

     public static void Main()
     {
          Application.Run(new ExplorerLike());
     }
     public ExplorerLike()
     {
          Text = "Windows Explorer-Like Program";
          BackColor = SystemColors.Window;
          ForeColor = SystemColors.WindowText;

               // Create controls.

          filelist        = new FileListView();
          filelist.Parent = this;
          filelist.Dock   = DockStyle.Fill;

          Splitter split  = new Splitter();
          split.Parent    = this;
          split.Dock      = DockStyle.Left;
          split.BackColor = SystemColors.Control;

          dirtree        = new DirectoryTreeView();
          dirtree.Parent = this;
          dirtree.Dock   = DockStyle.Left;
          dirtree.AfterSelect += 
                    new TreeViewEventHandler(DirectoryTreeViewOnAfterSelect);

               // Create View menu.

          Menu = new MainMenu();
          Menu.MenuItems.Add("&View");

          string[] astrView = { "Lar&ge Icons", "S&mall Icons", 
                                "&List", "&Details" };
          View[] aview = { View.LargeIcon, View.SmallIcon, 
                           View.List, View.Details }; 
          EventHandler eh = new EventHandler(MenuOnView);

          for (int i = 0; i < 4; i++)
          {
               MenuItemView miv = new MenuItemView();
               miv.Text = astrView[i];
               miv.View = aview[i];
               miv.RadioCheck = true;
               miv.Click += eh;

               if (i == 3)         // Default == View.Details
               {
                    mivChecked = miv;
                    mivChecked.Checked = true;
                    filelist.View = mivChecked.View;
               }
               Menu.MenuItems[0].MenuItems.Add(miv);
          }
          Menu.MenuItems[0].MenuItems.Add("-");

               // View Refresh menu item

          MenuItem mi = new MenuItem("&Refresh", 
                              new EventHandler(MenuOnRefresh), Shortcut.F5);
          Menu.MenuItems[0].MenuItems.Add(mi);
     }
     void DirectoryTreeViewOnAfterSelect(object obj, TreeViewEventArgs tvea)
     {
          filelist.ShowFiles(tvea.Node.FullPath);
     }
     void MenuOnView(object obj, EventArgs ea)
     {
          mivChecked.Checked = false;
          mivChecked = (MenuItemView) obj;
          mivChecked.Checked = true;

          filelist.View = mivChecked.View;
     }
     void MenuOnRefresh(object obj, EventArgs ea)
     {
          dirtree.RefreshTree();
     }
}
class MenuItemView: MenuItem
{
     public View View;
}

File: win-prog/ProgWinVB/Tree View and List View/ExplorerLike/ExplorerLike.vb
'---------------------------------------------
' ExplorerLike.vb (c) 2002 by Charles Petzold
'---------------------------------------------
Imports System
Imports System.Drawing
Imports System.Windows.Forms

Class ExplorerLike
    Inherits Form

    Private filelist As FileListView
    Private dirtree As DirectoryTreeView
    Private mivChecked As MenuItemView

    Shared Sub Main()
        Application.Run(New ExplorerLike())
    End Sub

    Sub New()
        Text = "Windows Explorer-Like Program"
        BackColor = SystemColors.Window
        ForeColor = SystemColors.WindowText

        ' Create controls.
        filelist = New FileListView()
        filelist.Parent = Me
        filelist.Dock = DockStyle.Fill

        Dim split As New Splitter()
        split.Parent = Me
        split.Dock = DockStyle.Left
        split.BackColor = SystemColors.Control

        dirtree = New DirectoryTreeView()
        dirtree.Parent = Me
        dirtree.Dock = DockStyle.Left
        AddHandler dirtree.AfterSelect, AddressOf DirectoryTreeViewOnAfterSelect

        ' Create View menu.
        Menu = New MainMenu()
        Menu.MenuItems.Add("&View")

        Dim astrView() As String = {"Lar&ge Icons", "S&mall Icons", _
                                    "&List", "&Details"}
        Dim aview() As View = {View.LargeIcon, View.SmallIcon, _
                               View.List, View.Details}
        Dim eh As EventHandler = AddressOf MenuOnView
        Dim i As Integer

        For i = 0 To 3
            Dim miv As New MenuItemView()
            miv.Text = astrView(i)
            miv.vu = aview(i)
            miv.RadioCheck = True
            AddHandler miv.Click, eh
            If i = 3 Then
                mivChecked = miv
                mivChecked.Checked = True
                filelist.View = mivChecked.vu
            End If
            Menu.MenuItems(0).MenuItems.Add(miv)
        Next i

        Menu.MenuItems(0).MenuItems.Add("-")

        ' View Refresh menu item
        Dim mi As New MenuItem("&Refresh", _
                        AddressOf MenuOnRefresh, Shortcut.F5)
        Menu.MenuItems(0).MenuItems.Add(mi)
    End Sub

    Private Sub DirectoryTreeViewOnAfterSelect(ByVal obj As Object, ByVal tvea As TreeViewEventArgs)
        filelist.ShowFiles(tvea.Node.FullPath)
    End Sub

    Private Sub MenuOnView(ByVal obj As Object, ByVal ea As EventArgs)
        mivChecked.Checked = False
        mivChecked = DirectCast(obj, MenuItemView)
        mivChecked.Checked = True
        filelist.View = mivChecked.vu
    End Sub

    Private Sub MenuOnRefresh(ByVal obj As Object, ByVal ea As EventArgs)
        dirtree.RefreshTree()
    End Sub
End Class

Class MenuItemView
    Inherits MenuItem

    Public vu As View
End Class
